pkg/loop/internal/pb/chain_reader.proto

```diff
@@ -34,3 +62,133 @@ message BoundContract {
   string name = 2;
   bool pending = 3;
 }
+
+enum ComparisonOperator {
+  Eq = 0;
+  Neq = 1;
+  Gt = 2;
+  Lt = 3;
+  Gte = 4;
+  Lte = 5;
+}
+
+message QueryKeyFilter {
+  string key = 1;
+  // Expressions are lightweight orm like DSL defined for filtering over common blockchain primitives.
+  repeated Expression expression = 2;
+}
+
+// Expression encapsulates a single unit of filtering logic, which can be a common blockchain primitive or a composite of boolean expressions. 
+// This allows for both simple and more complex nested expressions.
+message Expression {
+  oneof evaluator {
+    Primitive primitive = 1;
```

Comment:
Does it mean there will be a separate struct for filtering by topics/data words etc?

Reply:
That is correct. An example of filtering by data words:

```
type eventByWordFilter struct {
	EventSig       common.Hash
	WordIndex      uint8
	ValueComparers []primitives.ValueComparator
}

func (f *eventByWordFilter) Accept(visitor primitives.Visitor) {
	switch v := visitor.(type) {
	case *pgDSLParser:
		v.VisitEventByWordFilter(f)
	}
}
```

More detail is in the EVM implementation in the `chainlink` repo.